dynamic_programming/levenshtein_distance.nim

```diff
@@ -0,0 +1,72 @@
+## Levenshtein distance
+
+import tables
+
+
+func initSubsolutions(lenA: Natural, lenB: Natural): Table[(Natural, Natural), Natural] =
+  for indA in 0..lenA:
+    result[(Natural(indA), Natural(0))] = Natural(indA)
+
+  for indB in 1..lenB:
+    result[(Natural(0), Natural(indB))] = Natural(indB)
+
+
+func toKey(indA: int, indB: int): (Natural, Natural) =
+  return (Natural(indA), Natural(indB))
+
+
+func computeLevenshteinDistanceMatrix(strA: string, strB: string): Table[(
+    Natural, Natural), Natural] =
+  let
+    lenA = strA.len
+    lenB = strB.len
+
+  result = initSubsolutions(lenA, lenB)
+
+  for indA in 1..lenA:
+    for indB in 1..lenB:
+      let substitutionCost = (if strA[indA-1] == strB[indB-1]: 0 else: 1)
+      result[toKey(indA, indB)] = min(
+          [result[toKey(indA-1, indB)] + 1,
+           result[toKey(indA, indB-1)] + 1,
+           result[toKey(indA-1, indB-1)] + substitutionCost])
+
+
+func levenshteinDistance*(strA: string, strB: string): Natural =
+  ## Returns the
+  ## [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance)
+  ## between the input strings.
+  ## This implementation utilises dynamic programming.
+  runnableExamples:
+    doAssert levenshteinDistance("abc", "aXcY") == 2
+  return computeLevenshteinDistanceMatrix(strA, strB)[toKey(strA.len, strB.len)]
+
+
+when isMainModule:
+  import std/unittest
+
+  suite "levenshteinDistance":
+
+    type
+      TestCase = object
+        strA*, strB*: string
+        distance*: Natural
+
+    const testCases = @[
+      TestCase(strA: "", strB: "", distance: Natural(0)),
+      TestCase(strA: "", strB: "a", distance: Natural(1)),
+      TestCase(strA: "a", strB: "a", distance: Natural(0)),
+      TestCase(strA: "a", strB: "ab", distance: Natural(1)),
+      TestCase(strA: "saturday", strB: "sunday", distance: Natural(3)),
+      TestCase(strA: "kitten", strB: "sitting", distance: Natural(3)),
+      TestCase(strA: "abcdefghif", strB: "abCdeFghIfX", distance: Natural(4)),
+      TestCase(strA: "abc", strB: "aXcY", distance: Natural(2)),
+      ]
+
+    test "returns expected result":
+      for tc in testCases:
+        check tc.distance == levenshteinDistance(tc.strA, tc.strB)
+
+    test "is symmetric":
+      for tc in testCases:
+        check tc.distance == levenshteinDistance(tc.strB, tc.strA)
```

Comment:
```suggestion
  suite "Levenshtein Distance":

    const testCases = [
      (a: "", b: "", distance: 0),
      (a: "", b: "a", distance: 1),
      (a: "a", b: "a", distance: 0),
      (a: "a", b: "ab", distance: 1),
      (a: "saturday", b: "sunday", distance: 3),
      (a: "kitten", b: "sitting", distance: 3),
      (a: "abcdefghif", b: "abCdeFghIfX", distance: 4),
      (a: "abc", b: "aXcY", distance: 2),
      ]

    test "returns expected result":
      for (a, b, distance) in testCases:
        check levenshteinDistance(a, b) == distance.Natural

    test "is symmetric":
      for (a, b, distance) in testCases:
        check levenshteinDistance(a, b) == distance.Natural
```
1. `suite` is a user-facing string, not an identifier. No need for camelCasing it.
2. Ad-hoc bundling data together is generally a good idea, but object is clearly an overshoot here. Just an anonymous named tuple + destructuring works even better in this case, IMO.
3. `testCases` is supposed to be entered by hand so this is why I moved the Natural conversion to the `check` lines.

4. The main issue here (which I leave for you to implement if you find my reasoning compelling), is not distinguishing between different test cases and not utilising named `test` blocks for this. If one of the test cases fails it requires additional work to understand the issue. It would be nice to have the different test cases named. E.g.:
`empty strings`, `plus 1 character`, `minus 1 character`, `partial uppercasing`, etc.

Reply:
1. I used `levenshteinDistance` because, if this suite would fail, it is easier to find the function (most likely) causing problems. However, I do not have a very strong opinion here.
2. tbd
3. I made `distance` in `TestCase` a `Natural`, because it it a _requirement_ (i.e. we do not only want the result to be _some value_ but also to be a `Natural`). And this also simplifies the logic inside the `check`.
4. Do you know a good way how to express _parameterized tests_ in `nim`?

Reply:
Please have a look at ed9eabd2701210eda369ca9ddd98e9312df2767d. It is a pity that `check` does not support the syntax like `check, condition, message to display then fails`. At the moment I would not do anything super fancy and simply wait for some test framework for `nim`.

Please mark as _resolved_, if you agree that it is.

Reply:
Sorry, I'm not sure I'm done, but I opened a new thread as this is outdated.